### PR TITLE
WIP: vt6/term and vt6/mono modules

### PR DESCRIPTION
**Pull request by majewsky**
This is based on an earlier draft (#14) where monospace layouting was a capability within vt6/term. I chose to make an entirely new branch because so much has changed in vt6/core1.0 recently that stuff is way cleaner this way.

@laerling FYI - If you want to read this, start with `term1.0` and then continue with `mono1.0`. The latter depends on the former.

### CODE REVIEW

File: spec/term/1.0.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+A **terminal** is a triple of:
+- a Unicode string (the **terminal document**),
+- a **cursor** that points either at the start of the string, at its end, or between two characters, and
+- an input stream from which Unicode characters can be read.
+
+*Rationale:* A sole printer can provide such a terminal even though it has no input, because the definition does not require that the input stream ever becomes ready for reading.
+
+### 1.2. Standard output
+
+When a client writes a byte string into its standard output, the server SHALL assume that this byte string is encoded in [UTF-8](https://tools.ietf.org/html/rfc3629), and decode it into a string of Unicode characters.
```
Comment (majewsky) [inline spec/term/1.0.md]: > When a client writes a byte string into its standard output

That stdout may not be connected to the terminal.

File: spec/term/1.0.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+**This is a non-normative draft.**
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+This document uses the predefined property types from [section 2.4 of `vt6/core1.0`](https://vt6.io/std/core/1.0/#section-2-4).
+
+## 1. Scope
+
+This module defines how a terminal works, how clients' standard output is reflected on the terminal, and how user input is presented on clients' standard input.
+Besides [`vt6/core`](https://vt6.io/std/core/), this is the only module that we expect to be available in **every** VT6 server implementation across the entire feature spectrum, from graphical terminal emulators all the way down to continuous stationery printers.
```
Comment (laerling) [inline spec/term/1.0.md]: term module: Can we safely allow vt6 client implementers to assume the term module to always be present? Or should we advise them to always `want` it?

File: spec/term/1.0.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+A **Unicode document** is a string of Unicode characters with a cursor.
+The cursor points either at the start of the string (before the first character), at its end (after the last character) or between two adjacent characters.
+
+When a VT6 server agrees to using this module, it MUST have a Unicode document (the **terminal document**) that its clients can interact with as described in section 1.2.
+All clients MUST act upon the same terminal document.
+
+If the server takes text input from a user or some other source, it MAY pass this input to the clients as described in section 1.3.
+
+*Rationale:* This is a "MAY" because some servers (e.g. continuous stationery printers) do not have an input method.
```
Comment (laerling) [inline spec/term/1.0.md]: Nitpick: That's what the "If" is for. I don't know why there is a "MAY". I SHOULD be a "MUST" ;)

File: spec/term/1.0.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+*Rationale:* Clients should be able to rely on the server interpreting their standard output faithfully.
+
+### 1.2. Standard output
+
+When the server receives a byte string written by a client into its standard output, the server SHALL assume that this byte string is encoded in [UTF-8](https://tools.ietf.org/html/rfc3629), and decode it into a string of Unicode characters.
+Invalid byte sequences or invalid code points SHALL be handled gracefully, for instance, by replacing them with the replacement character (U+FFFD).
+
+When the `term.output-protected` property is true, the server MUST then discard all characters from this string that are in the general categories Cc ("Other, control"), Cf ("Other, format"), Cs ("Other, surrogate"), Co ("Other, private use") and Cn ("Other, not assigned"), except for U+0009 (Horizontal Tab), U+000A (New Line), U+000C (Form Feed) and U+000D (Carriage Return).
+These characters are called **active characters** for the remainder of this specification.
```
Comment (laerling) [inline spec/term/1.0.md]: Not entirely clear whether "These characters" refers to "all characters from this string that are in the general categories" or "except for".

File: spec/term/1.0.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+Characters that are in the general category Co ("Other, private use") MAY be exempted from the treatment described above if the characters in question have a graphic representation in the font that the terminal uses to display its character array.
+
+Finally, the server MUST discard each U+000A (new line) character that immediately succeeds a U+000D (carriage return) character, and then convert each U+000D character into a U+000A character.
+
+*Rationale:* This rule ensures consistent line break handling across different platforms, i.e., CR+NL = CR = NL.
+
+The remaining string of Unicode characters SHALL then be inserted into the terminal document at the position pointed to by the cursor.
+The cursor SHALL be updated to point at the position directly after the inserted string.
+If the terminal document would grow too large to fit within the program's memory because of this operation, the server MAY choose to discard some of the text at the start of the document.
```
Comment (laerling) [inline spec/term/1.0.md]: Practically the same as in 1.5.

File: spec/term/1.0.md (hunk 2 above)
Comment (laerling) [inline spec/term/1.0.md]: Except muxers and proxies...

File: spec/term/1.0.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+If the cursor points into that prefix or at its start, the cursor SHALL be moved to the end of the prefix before the prefix is discarded.
+
+The server SHOULD choose the prefix such that its end is a line separator, paragraph separator, or a line break control character (U+0010).
+
+*Rationale:* In other words, servers should prefer to discard whole lines at a time.
+Chomping off only parts of a line causes unnecessary confusion on the side of the user.
+
+## 2. Message types for `vt6/term1`
+
+### 2.1. The `term.make-stdio` message
```
Comment (laerling) [inline spec/term/1.0.md]: `make-stdio` and `is-stdio` should not be part of `term`, but of `core`, since muxers and proxies must also be able to turn a connection into stdio.

File: spec/term/1.0.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+- Directionality: server to client
+- Number of arguments: zero
+
+A server sends this message as a successful response to a `term.make-stdio` message.
+After this message, this server connection ceases to be used as a server connection and can be used as a standard input/output instead:
+
+TODO explain in detail (how stdin works, how stdout works, how anything that might send a server-initiated message is cancelled)
+
+TODO maybe remove the notion of standard input/output from core1.0 entirely and introduce it here?
```
Comment (laerling) [inline spec/term/1.0.md]: See comment above.

File: spec/term/1.0.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+## 3. Properties for `vt6/term1`
+
+TODO define connection-lifetime-scoped properties
+
+### 3.1. The `term.width` property
+
+- Acceptable values: unsigned integers
+- Can be set by client: no
+
+The value of this property is the maximum number of printable ASCII characters (Unicode codepoints U+0020-U+007F) that can be displayed on the terminal viewport without having to break the line or overflow the viewport.
```
Comment (laerling) [inline spec/term/1.0.md:112]: I don't see how this makes any sense outside of mono.

File: spec/term/1.0.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+It MAY be computed as follows:
+
+1. For any printable ASCII codepoint `c`, let `N(c)` be the maximum length of a string containing only the codepoint `c`, such that a rendering of this string fits in the width of the terminal viewport without line breaks or overflows.
+2. The value of the `term.width` property is `N(c')` where `c'` is that value of `c` that minimizes `N(c)`.
+
+*Rationale:* This is a "MAY" because servers may employ more sophisticated computations to account for ligatures, or simpler computations if they use monospace fonts.
+
+The value of this property SHALL NOT be zero.
+If the computation above would result in a value of zero, the server SHALL report a value of one SHALL instead.
```
Comment (laerling) [inline spec/term/1.0.md]: What is a SHALL? I suppose you mean "the value of one"?

File: spec/term/1.0.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+- Can be set by clients: no
+
+A value of true indicates that the server reflows the entire terminal document when the viewport width (i.e., the value of the `term.viewport-width` property) changes.
+
+A value of false indicates that the server MAY, after a viewport width change, continue to display unchanged parts of the terminal document according to the old viewport width.
+
+### 3.8. The `term.output-wordwrap` property
+
+- Acceptable values: booleans
+- Can be set by clients: no
```
Comment (laerling) [inline spec/term/1.0.md:184]: I don't see why we should forbid clients to set this.

File: spec/term/1.0.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+### 3.1. The `term.width` property
+
+- Acceptable values: unsigned integers
+- Can be set by client: no
+
+The value of this property is the maximum number of printable ASCII characters (Unicode codepoints U+0020-U+007F) that can be displayed on the terminal viewport without having to break the line or overflow the viewport.
+
+It MAY be computed as follows:
+
+1. For any printable ASCII codepoint `c`, let `N(c)` be the maximum length of a string containing only the codepoint `c`, such that a rendering of this string fits in the width of the terminal viewport without line breaks or overflows.
```
Comment (laerling) [inline spec/term/1.0.md:116]: One codepoint? I'm not sure what a codepoint is anymore.

File: spec/mono/1.0.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+In fact, VT6-compliant terminals may not offer monospace layouting at all.
+
+### 1.1. Dependencies to other modules
+
+VT6 servers MUST NOT agree to using this module unless they have already agreed to using any version of `vt6/term` on the same server connection.
+
+### 1.2. Monospace layouting
+
+We take the term **terminal** to refer to a terminal as defined in [`vt6/term1.0`](https://vt6.io/std/term/1.0/).
+When the value of the `mono.enabled` property is changed, the server SHALL append a paragraph separator (U+2029) to the terminal document, and move the terminal's cursor to point at the end of the terminal document, after that paragraph separator.
```
Comment (laerling) [inline spec/mono/1.0.md]: What's the behavior of form feed compared to paragraph separator? Could using form feed be reasonable as well or even instead?

File: spec/mono/1.0.md [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+After this operation, the terminal's cursor SHALL NOT move past this paragraph separator, even when the specification of another module would allow it.
```
Comment (laerling) [inline spec/mono/1.0.md:33]: What about text selection by mouse? E. g. when I want to reuse a text snippet of non-mono output again in a new mono-'environment'? Does mouse selection not imply setting the cursor?

File: spec/mono/1.0.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+- Acceptable values: booleans
+- Can be set by client: yes
+
+The value of this property indicates whether the terminal is in monospace layouting mode, as described in section 1.2.
+
+When this property is set by a client using a `core.set` message, and its server connection is then closed, the property SHALL be reset to the previous value, unless the property's value has also been set on some other server connection after that `core.set` message.
+
+*Rationale:* When a client sets `mono.enabled`, prints some text and exits, the shell will then observe the previous value of `mono.enabled` when printing its prompt.
+
+TODO consider upstreaming this scoping logic into core
```
Comment (laerling) [inline spec/mono/1.0.md]: Could it occur that two or more programs inside a single pipe (or just two or more programs that run at the same time on the same frame) compete in a way that e. g. one sets `mono.enabled` to true and expects it to be true until set otherwise, but the other one sets `mono.enabled` to false while the first one still expects it to be true?
This is not only a problem for `mono.enabled`, but is necessary to consider when upstreaming to core.

**Review by laerling — CHANGES_REQUESTED** [review]

File: spec/term/1.0.md (hunk 3 above)
Comment (majewsky) [reply spec/term/1.0.md]: No. Not all text input that a terminal takes from a user must be passed on to clients' standard inputs.

File: spec/term/1.0.md (hunk 2 above)
Comment (majewsky) [reply spec/term/1.0.md]: Muxers and proxies still need to implement `term`. They *are not* the terminal, but they have an internal model of a terminal that they present towards their own clients. A muxer/proxy that only has `core` is less than useless.

File: spec/term/1.0.md (hunk 6 above)
Comment (majewsky) [reply spec/term/1.0.md]: See above.

File: spec/term/1.0.md (hunk 7 above)
Comment (majewsky) [reply spec/term/1.0.md]: See rebuttal above.

File: spec/term/1.0.md (hunk 3 above)
Comment (laerling) [reply spec/term/1.0.md]: Alright, but then that is the reason which should appear in the text, not "some servers do not have an input method".

File: spec/term/1.0.md (hunk 10 above)
Comment (majewsky) [reply spec/term/1.0.md:184]: If this can be set by a client, we have to define how the layout works when fundamental layout properties can change over time. I do not find this a productive use of our time since I don't see a usecase for a client changing this property.

File: spec/term/1.0.md (hunk 11 above)
Comment (majewsky) [reply spec/term/1.0.md:116]: Yeah, me neither... :sob:

File: spec/term/1.0.md (hunk 2 above)
Comment (laerling) [reply spec/term/1.0.md]: So proxies and muxers would just forward all term-related messages from the terminal to the client and not modify it. Makes sense.

File: spec/mono/1.0.md (hunk 12 above)
Comment (majewsky) [reply spec/mono/1.0.md]: A form feed could trigger pagination on printers. I don't find it helpful to have page breaks between each single shell command and prompt (in the extreme case that the shell does not set `mono.enabled = true`, but every client run by it does).

File: spec/mono/1.0.md (hunk 13 above)
Comment (majewsky) [reply spec/mono/1.0.md:33]: This specification does not concern a cursor operated by the user. The cursor we're talking about here is denoting the place where additional stdout gets inserted into the terminal document.

File: spec/mono/1.0.md (hunk 14 above)
Comment (majewsky) [reply spec/mono/1.0.md]: Yes, there is a race condition here, but I don't think it's important in practice. `mono.enabled` defaults to `false`, so most clients that set it are going to set it to true.

File: spec/term/1.0.md (hunk 6 above)
Comment (laerling) [reply spec/term/1.0.md]: Let me rephrase: `make-stdio` and `is-stdio` should not be part of `term`, but of `core`, since those are functionalities that *every* server must support, regardless of whether it `have`s `term` or not.
I know, all servers will `have term`, but the point is that the `make-stdio` and `is-stdio` are independent of that.

We must not define things that every server must have inside a non-`core` module that every server must have, when those things don't have anything to do with it semantically!

File: spec/term/1.0.md (hunk 2 above)
Comment (majewsky) [reply spec/term/1.0.md]: Well, not "not modify it". The only modification they do is multiplexing it onto a single connection.

File: spec/term/1.0.md (hunk 10 above)
Comment (laerling) [reply spec/term/1.0.md:184]: Accepted.

File: spec/term/1.0.md (hunk 5 above)
Comment (majewsky) [reply spec/term/1.0.md]: Fixed by folding 1.5 into this part.

File: spec/term/1.0.md (hunk 11 above)
Comment (laerling) [reply spec/term/1.0.md:116]: No seriously, let's talk about this. I don't see why you calculate a width of a whole (virtual) line by one single codepoint.

File: spec/term/1.0.md (hunk 2 above)
Comment (laerling) [reply spec/term/1.0.md]: That's what I meant.

File: spec/mono/1.0.md (hunk 14 above)
Comment (laerling) [reply spec/mono/1.0.md]: But the race condition could be bad the other way round, as in that a client expects the terminal to be fully unicode-capable while another client already set `mono.enabled` to true a long time ago.

File: spec/term/1.0.md (hunk 7 above)
Comment (laerling) [reply spec/term/1.0.md]: See my ranty response above.

File: spec/term/1.0.md (hunk 6 above)
Comment (majewsky) [reply spec/term/1.0.md]: Agreed. Move this to `core` to enforce a logical separation between `core` (which corresponds roughly to OSI layers 1-4) and `term` (which corresponds roughly to OSI layers 5-7).

File: spec/term/1.0.md (hunk 8 above)
Comment (majewsky) [reply spec/term/1.0.md:112]: OK.

File: spec/mono/1.0.md (hunk 14 above)
Comment (majewsky) [reply spec/mono/1.0.md]: After short discussion, the actual problem here is that it should not be possible for a client to set `mono.enabled` *unless* their stdout is connected to the terminal. So this question ties into #19.

File: spec/term/1.0.md (hunk 11 above)
Comment (majewsky) [reply spec/term/1.0.md:116]: Will be resolved when this property moves into `mono` since the definition can accordingly be simplified.

**Comment by majewsky** [comment]
@laerling I rebased this and adjusted the wording to match #39. There are still plenty of TODOs that I will address at a later time.
